lib/test_hook.rb

```diff
@@ -76,18 +85,75 @@ def post_process_datasets(datasets)
     end
   end
 
-  # Split query by '-- DATASET' line match
-  # First match is teacher's solution
-  # Rest are Datasets
-  # Example:
-  #   select * from table
-  #   -- DATASET
-  #   insert into 1
-  #   -- dataset
-  #   insert into 2
-  def parse_test(content)
-    query = content.split(/\s*--\s*dataset\s*\n+/i)
-    return query.shift, query
+  # Test should have one of these formats:
+  #
+  # Solution by query:
+  # { solution_type: 'query',
+  #   solution_query: 'SELECT * FROM ...',
+  #   examples: [
+  #     { data: "INSERT INTO..." }
+  #   ]
+  # }
+  #
+  # Solution by datasets:
+  # { solution_type: 'datasets',
+  #   examples: [
+  #     { data: "INSERT INTO...",
+  #       solution: "id|field\n1|row1..."
+  #     }
+  #   ]
+  # }
+  def parse_test(test)
+    test = OpenStruct.new (YAML.load(test).deep_symbolize_keys)
+    @solution_type = test.solution_type.to_sym
+
+    case @solution_type
```

Comment:
Since you are doing this case in two different places, I'd recommend reifying the polimorphic `query` and `datasets` objects, and implementing a `parse_solution_type` that performs this type check only once and returns the proper objects.

Reply:
Ok, this is code climate issue too. I'll fix it